### PR TITLE
SALTO-5819 Google WS - CV

### PR DESCRIPTION
**Pull request by Uri-Bar**
SALTO-5819 Google WS - CV
---

_Additional context for reviewer_

---
_Release Notes_: 
none

---
_User Notifications_: 
none

### CODE REVIEW

File: packages/google-workspace-adapter/src/change_validator.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import {
+  domainReadOnlyValidator,
+  roleReadOnlyValidator,
+  schemaFieldsValidator,
+  systemRoleValidator,
+  groupMemberRoleValidator,
+  roleAssignmentAdditionValidator,
+} from './change_validators'
+
+export default (): Record<string, ChangeValidator> => ({
```
Comment (netama) [inline packages/google-workspace-adapter/src/change_validator.ts:26]: **can be in a different PR** (and technically maybe part of SALTO-5572) - we also need to add the option to disable change validators in the config. there's a `changeValidatorNames` field you can pass into `createUserConfigType` in the wrapper, and then we should only run the validators that pass `createSkipParentsOfSkippedInstancesValidator` (see the zendesk [implementation](https://github.com/salto-io/salto/blob/main/packages/zendesk-adapter/src/change_validator.ts#L201-L204) for example). I think we should have `createSkipParentsOfSkippedInstancesValidator` in the wrapper's adapter.ts (and also pass in the validator keys so they're available in the config). lmk if you want to discuss more


also (in this PR if easy) - let's add a version of this file (with a TODO of where to add validators) to the serviceplaceholder adapter?

**Review by netama — COMMENTED** [review]
looks good and well thought out :) added some comments (some not for this PR), the main one I'm not sure I understand is the role assignment validator

File: packages/google-workspace-adapter/src/change_validators/role_assignment_addition.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  }
+  return false
+}
+
+export const roleAssignmentAdditionValidator: ChangeValidator = async (changes, elementSource) => {
+  if (elementSource === undefined) {
+    return []
+  }
+  return awu(changes)
+    .filter(isAdditionChange)
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/google-workspace-adapter/src/change_validators/role_assignment_addition.ts:46]: yes

File: packages/google-workspace-adapter/src/change_validators/role_assignment_addition.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import { ROLE_ASSIGNMENT_TYPE_NAME } from '../constants'
+
+const { awu } = collections.asynciterable
+
+const isSecurityGroup = async (instance: InstanceElement, elementSource: ReadOnlyElementsSource): Promise<boolean> => {
+  const groupReference = instance.value.assignedTo
+  if (isReferenceExpression(groupReference)) {
+    const group = await groupReference.getResolvedValue(elementSource)
+    const labels = Object.keys(group.value.labels)
+    return !labels.includes('cloudidentity_googleapis_com_groups_security@vvdv')
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/google-workspace-adapter/src/change_validators/role_assignment_addition.ts:36]: 1. I changed the name to isNonSecurityGroup
2. just nacl-casing a key returned from the service. I know we can do here some work in order for it to be cleaner but it is ok for now IMO

File: packages/google-workspace-adapter/src/change_validators/role_assignment_addition.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    .filter(isInstanceElement)
+    .filter(instance => instance.elemID.typeName === ROLE_ASSIGNMENT_TYPE_NAME)
+    .filter(async instance => isSecurityGroup(instance, elementSource))
+    .map(
+      ({ elemID }): ChangeError => ({
+        elemID,
+        severity: 'Error',
+        message: 'Can not create role assignment for non security groups',
+        detailedMessage: 'Can not create role assignment for non security groups',
+      }),
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/google-workspace-adapter/src/change_validators/role_assignment_addition.ts:57]: changed the name to isNonSecurityGroup

File: packages/google-workspace-adapter/src/change_validators/role_read_only.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    .flatMap(instance => [
+      {
+        elemID: instance.elemID,
+        severity: 'Info',
+        message: 'Can not edit isSuperAdminRole trough the API',
+        detailedMessage: `Can not edit ${instance.value.roleName} isSuperAdminRole field trough the API, we will change it to false`,
+      },
+    ])
+
+  const modificationChangeErrors: ChangeError[] = changes
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/google-workspace-adapter/src/change_validators/role_read_only.ts:53]: I basically agree, I just wanted to be grateful in the creation as it copy the whole instance.
I think we can face it later on, plus the `isSystemRole` already has his own change validator that will throw an error if some change it

File: packages/google-workspace-adapter/src/change_validators/system_role.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    .map(getChangeData)
+    .filter(isInstanceElement)
+    .filter(instance => instance.elemID.typeName === ROLE_TYPE_NAME)
+    .filter(instance => instance.value.isSystemRole === true)
+    .flatMap(instance => [
+      {
+        elemID: instance.elemID,
+        severity: 'Error',
+        message: 'Can not edit system roles trough the API',
+        detailedMessage: `Can not edit the ${instance.value.roleName} system roles trough the API`,
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/google-workspace-adapter/src/change_validators/system_role.ts:30]: will remove

File: packages/google-workspace-adapter/src/change_validator.ts (hunk 1 above)
Comment (Uri-Bar) [reply packages/google-workspace-adapter/src/change_validator.ts:26]: will open a ticket and will add to serviceplaceholder

**Review by netama — APPROVED** [review]
LGTM 🎈 @liorn fyi for when you're back in case you want to adjust any of the messages

**Pull request merged**